### ISSUE TITLE
Missing fields result in incorrect calculations in other series when using GROUP BY

### ISSUE THREAD

**Issue by britcey** [influxdata/influxdb#7895]
### Bug report

Background: loading SNMP data into Influx and calculating interface utilization for each interface.  Percentages are off by orders of magnitude in some cases, apparently due to missing fields _in other series_.  Specifying tags for device & interface produce correct results.

__System info:__ 
InfluxDB 1.2, Oracle Linux 5 & Mac OS Sierra

__Steps to reproduce:__

(bear with me, it has taken many, many hours to get to this point).

1. Load test data (real, anonymized data):

[test_data.txt.gz](https://github.com/influxdata/influxdb/files/734016/test_data.txt.gz)

```
$ influx -precision s -import -path ./test_data.txt.gz -compressed
```
Will create a 'pct_calc_test' database with a 'interfaceStats' measurement.

2. Calculate interface utilization using the derivative of the in/out counters divided by the interface's' bandwidth (ifHighSpeed * 125,000 = bandwidth in bytes), group by device and interface.
```
$ influx -precision rfc3339 -database pct_calc_test

> SELECT non_negative_derivative(mean(ifHCInOctets), 1s) / (last(ifHighSpeed) * 125000) * 100 AS rx_pct, non_negative_derivative(mean(ifHCOutOctets), 1s) / (last(ifHighSpeed) * 125000) * 100 AS tx_pct INTO interfaceRates FROM interfaceStats where time > '2017-01-26T21:30:00Z' and time < '2017-01-26T22:30:00Z' GROUP BY time(10m), agent_host, ifDescr
```

This results in incorrect percentages - easiest to catch where it's > 100%:

```
> select * from interfaceRates where (tx_pct > 100 or rx_pct > 100)
name: interfaceRates
time                 agent_host          ifDescr               rx_pct             tx_pct
----                 ----------          -------               ------             ------
2017-01-26T22:00:00Z 126ba64.example.com Ethernet1/1           13.24685117779948  699.6494319111979
2017-01-26T22:00:00Z c23005e.example.com GigabitEthernet1/0/1  127.10156508893229 6.441728533333333
2017-01-26T22:00:00Z 49084e9.example.com GigabitEthernet1/0/40 134.6534640222656  98.39347193333333
2017-01-26T22:00:00Z 9de351f.example.com Port-channel17        799.406482         571.3652787333333
2017-01-26T22:00:00Z 1b44a18.example.com Port-channel11        38.1950939109375   301.27599144479166
2017-01-26T22:10:00Z 9de351f.example.com Port-channel17        635.877054         454.43473373333336
2017-01-26T22:10:00Z 1b44a18.example.com Port-channel11        33.064395155729166 258.05897475520834
2017-01-26T22:10:00Z 126ba64.example.com Ethernet1/1           10.692711022200521 564.7613418221354
2017-01-26T22:10:00Z 49084e9.example.com GigabitEthernet1/0/40 102.27434711106771 67.96335253333334
```

3. Repeat the query, but specify the device (agent_host) and interface (ifDescr) for one of the above results:
```
> SELECT non_negative_derivative(mean(ifHCInOctets), 1s) / (last(ifHighSpeed) * 125000) * 100 AS rx_pct, non_negative_derivative(mean(ifHCOutOctets), 1s) / (last(ifHighSpeed) * 125000) * 100 AS tx_pct FROM interfaceStats where agent_host='49084e9.example.com' and ifDescr='GigabitEthernet1/0/40' and time > '2017-01-26T21:30:00Z' and time < '2017-01-26T22:30:00Z' GROUP BY time(10m), agent_host, ifDescr
name: interfaceStats
tags: agent_host=49084e9.example.com, ifDescr=GigabitEthernet1/0/40
time                 rx_pct             tx_pct
----                 ------             ------
2017-01-26T21:30:00Z
2017-01-26T21:40:00Z
2017-01-26T21:50:00Z
2017-01-26T22:00:00Z 1.3465346402226563 0.9839347193333334
2017-01-26T22:10:00Z 1.022743471110677  0.6796335253333333
2017-01-26T22:20:00Z
```
Those percentages are correct.

4. Examine the raw data - nothing is amiss:
```
> SELECT * FROM interfaceStats where agent_host='49084e9.example.com' and ifDescr='GigabitEthernet1/0/40' and time > '2017-01-26T21:30:00Z' and time < '2017-01-26T22:30:00Z'
name: interfaceStats
time                 agent_host          ifAdminStatus ifDescr               ifHCInBroadcastPkts ifHCInOctets  ifHCOutBroadcastPkts ifHCOutOctets ifHighSpeed ifInDiscards ifInErrors ifOperStatus ifOutDiscards ifOutErrors ifType
----                 ----------          ------------- -------               ------------------- ------------  -------------------- ------------- ----------- ------------ ---------- ------------ ------------- ----------- ------
2017-01-26T21:55:06Z 49084e9.example.com 1             GigabitEthernet1/0/40 50301912            6733772566804 711609407            3218765650890 1000        0            0          1            1375          0           6
2017-01-26T21:58:05Z 49084e9.example.com 1             GigabitEthernet1/0/40 50303466            6733810602521 711621985            3218822456913 1000        0            0          1            1375          0           6
2017-01-26T22:01:06Z 49084e9.example.com 1             GigabitEthernet1/0/40 50304831            6734191133280 711634660            3219163176818 1000        0            0          1            1375          0           6
2017-01-26T22:04:06Z 49084e9.example.com 1             GigabitEthernet1/0/40 50306017            6734978345140 711646957            3219599852359 1000        0            0          1            1375          0           6
2017-01-26T22:07:06Z 49084e9.example.com 1             GigabitEthernet1/0/40 50307403            6735234978508 711658344            3219832985646 1000        0            0          1            1375          0           6
2017-01-26T22:10:05Z 49084e9.example.com 1             GigabitEthernet1/0/40 50308976            6735568543246 711670166            3220041730085 1000        0            0          1            1375          0           6
```

However, a different interface on that same host is missing some field data:
```
> SELECT * FROM interfaceStats where agent_host='49084e9.example.com' and ifDescr='GigabitEthernet1/1/2' and time > '2017-01-26T21:30:00Z' and time < '2017-01-26T22:30:00Z'
name: interfaceStats
time                 agent_host          ifAdminStatus ifDescr              ifHCInBroadcastPkts ifHCInOctets ifHCOutBroadcastPkts ifHCOutOctets ifHighSpeed ifInDiscards ifInErrors ifOperStatus ifOutDiscards ifOutErrors ifType
----                 ----------          ------------- -------              ------------------- ------------ -------------------- ------------- ----------- ------------ ---------- ------------ ------------- ----------- ------
2017-01-26T21:55:06Z 49084e9.example.com 1             GigabitEthernet1/1/2                                                                     10          0            0          2            0             0           6
2017-01-26T21:58:05Z 49084e9.example.com 1             GigabitEthernet1/1/2                                                                     10          0            0          2            0             0           6
2017-01-26T22:01:06Z 49084e9.example.com 1             GigabitEthernet1/1/2                                                                     10          0            0          2            0             0           6
2017-01-26T22:04:06Z 49084e9.example.com 1             GigabitEthernet1/1/2                                                                     10          0            0          2            0             0           6
2017-01-26T22:07:06Z 49084e9.example.com 1             GigabitEthernet1/1/2                                                                     10          0            0          2            0             0           6
2017-01-26T22:10:05Z 49084e9.example.com 1             GigabitEthernet1/1/2                                                                     10          0            0          2            0             0           6
```
This seems to be the cause of the issues.

5. Load version of the test data where missing fields values have been set to 0:

[test_data_filled.txt.gz](https://github.com/influxdata/influxdb/files/734015/test_data_filled.txt.gz)

```
$ influx -precision s -import -path ./test_data_filled.txt.gz -compressed
```

6. Calculate utilization and inspect (correct) data
```
$ influx -precision rfc3339 -database pct_calc_test_filled

> SELECT non_negative_derivative(mean(ifHCInOctets), 1s) / (last(ifHighSpeed) * 125000) * 100 AS rx_pct, non_negative_derivative(mean(ifHCOutOctets), 1s) / (last(ifHighSpeed) * 125000) * 100 AS tx_pct INTO interfaceRates FROM interfaceStats where time > '2017-01-26T21:30:00Z' and time < '2017-01-26T22:30:00Z' GROUP BY time(10m), agent_host, ifDescr

> select * from interfaceRates where (tx_pct > 100 or rx_pct > 100)

> select * from interfaceRates where agent_host='49084e9.example.com' and ifDescr='GigabitEthernet1/0/40'
name: interfaceRates
time                 agent_host          ifDescr               rx_pct             tx_pct
----                 ----------          -------               ------             ------
2017-01-26T22:00:00Z 49084e9.example.com GigabitEthernet1/0/40 1.3465346402226563 0.9839347193333334
2017-01-26T22:10:00Z 49084e9.example.com GigabitEthernet1/0/40 1.022743471110677  0.6796335253333333
```

__Expected behavior:__ 
Mass calculation using GROUP BY should produce the same results as when specifying tags.

__Actual behavior:__ 
Calculations are off by two orders of magnitude.

**Comment by e-dard** [comment]
@britcey thanks a lot for the detailed issue, it should help the team track down the issue.

**Comment by jsternberg** [comment]
It looks like this is because the math iterator only tries to match on time and doesn't try to match on name and tags since it's not expecting that. Because `non_negative_derivative()` emits one fewer point than `last()`, that seems to be where the issue is coming from.

I think I might have a fix, but this was a more complicated one. After I make a fix, would you be able to help test it by seeing if it fixes your issue?

**Comment by britcey** [comment]
Absolutely, though the sample data I included can be used to test as well.

**Comment by jsternberg** [comment]
I've submitted a PR in #7937. Can you try that out?

**Comment by britcey** [comment]
That appears to work correctly - using the steps defined above, the info in interfaceRates looks correct, e.g.:
```
> select * from interfaceRates where agent_host='49084e9.example.com' and ifDescr='GigabitEthernet1/0/40'
name: interfaceRates
time                 agent_host          ifDescr               rx_pct             tx_pct
----                 ----------          -------               ------             ------
2017-01-26T22:00:00Z 49084e9.example.com GigabitEthernet1/0/40 1.3465346402226563 0.9839347193333334
2017-01-26T22:10:00Z 49084e9.example.com GigabitEthernet1/0/40 1.022743471110677  0.6796335253333333
```

**Comment by jwilder** [comment]
Fixed via #7937
